mediatimestamp/immutable.py

```diff
@@ -0,0 +1,939 @@
+the operands
+
+TS + TO = TS
+TS + TS = TS (treats 2nd TS as TO)
+TO + TS = TS
+TO + TO = TO
+
+TS - TO = TS
+TS - TS = TO
+TO - TS = TO (treats TS as TO)
+TO - TO = TO
+
++= and -= always give the same result type as the first operand
+
+Time Offsets can always be multiplied or divided by integers and floats and always give TimeOffsets as a result.
+
+Timestamps multiplied by integers or floats will be treated as TimeOffsets
+"""
+
+from __future__ import print_function
+from __future__ import absolute_import
+
+from six import integer_types
+
+import calendar
+import time
+import re
+from datetime import datetime
+from dateutil import tz
+try:
+    import pyipputils.ipptimestamp
+    IPP_UTILS = True
+except ImportError:
+    IPP_UTILS = False
+
+__all__ = ["TimeOffset", "Timestamp", "TimeRange"]
+
+from .constants import MAX_NANOSEC, MAX_SECONDS, UTC_LEAP
+from .exceptions import TsValueError
+
+from . import BaseTimeOffset, BaseTimeRange
+
+
+def _parse_seconds_fraction(frac):
+    """ Parse the fraction part of a timestamp seconds, using maximum 9 digits
+    Returns the nanoseconds
+    """
+    ns = 0
+    mult = TimeOffset.MAX_NANOSEC
+    for c in frac:
+        if c < '0' or c > '9' or int(mult) < 1:
+            break
+        mult = mult / 10
+        ns += mult * int(c)
+    return ns
+
+
+def _parse_iso8601(iso8601):
+    """ Limited ISO 8601 timestamp parse; expands YYYY-MM-DDThh:mm:ss.s
+    Returns tuple of (year, month, day, hours, mins, seconds, nanoseconds)
+    """
+    iso_date_time = iso8601.split("T")
+    if len(iso_date_time) != 2:
+        raise TsValueError("invalid or unsupported ISO 8601 UTC format")
+    iso_date = iso_date_time[0].split("-")
+    iso_time = iso_date_time[1].split(":")
+    if len(iso_date) != 3 or len(iso_time) != 3:
+        raise TsValueError("invalid or unsupported ISO 8601 UTC format")
+    sec_frac = iso_time[2].split(".")
+    if len(sec_frac) != 1 and len(sec_frac) != 2:
+        raise TsValueError("invalid or unsupported ISO 8601 UTC format")
+    sec = sec_frac[0]
+    ns = 0
+    if len(sec_frac) > 1:
+        ns = _parse_seconds_fraction(sec_frac[1])
+    return (int(iso_date[0]), int(iso_date[1]), int(iso_date[2]), int(iso_time[0]), int(iso_time[1]), int(sec), ns)
+
+
+class TimeOffset(BaseTimeOffset):
+    """A nanosecond precision immutable time difference object.
+
+    Note that the canonical representation of a TimeOffset is seconds:nanoseconds, e.g. "4:500000000".
+    TimeOffsets in seconds.fractions format (e.g. "4.5") can be parsed, but should not be used for serialization or
+    storage due to difficulty disambiguating them from floats.
+    """
+    ROUND_DOWN = 0
+    ROUND_NEAREST = 1
+    ROUND_UP = 2
+
+    MAX_NANOSEC = MAX_NANOSEC
+    MAX_SECONDS = MAX_SECONDS
+
+    def __init__(self, sec=0, ns=0, sign=1):
+        (sec, ns, sign) = self._make_valid(int(sec), int(ns), int(sign))
```

Comment:
If I've read `_make_valid()` correctly, this will silently "fix" the timestamp you supplied in a way that's not immediately obvious from the documentation; `ns > MAX_NANOSEC` is clamped rather than converted, and sec < 0 with the wrong sign is clamped to zero. 
Perhaps that should be called out, or the logic from `__add__` (line 362) consolidated into `_make_valid()`?